File: exam/src/factorys/DocumentFactory.ts
import { 
    IBaseDocument, Contract, FinancialReport,
     Proposal, Document,
} from "../models";


export class DocumentFactory {
    public static create(type: string, baseDocument: IBaseDocument): Document {
        switch (type) {
            case "Contract":
                return new Contract(baseDocument);
            case "FinancialReport":
                return new FinancialReport(baseDocument);
            case "Proposal":
                return new Proposal(baseDocument);
            default:
                throw new Error(`Document type ${type} not supported`);
        }
    }
}

File: exam/src/fileProcessing/DocumentFactory.java
package fileProcessing;

import documents.*;

import java.util.HashMap;

public class DocumentFactory {
    public static Document create(DOC_TYPE type, String filename, HashMap<String, String> metadata) {
        switch (type) {
            case DOC_TYPE.Contract -> {
                return new Contract(filename, metadata);
            }
            case DOC_TYPE.Report ->  {
                return new Report(filename, metadata);
            }
            // Pondria otro case del switch, pero así me ahorro la excepción innecesaria!
            default -> {
                return new Proposal(filename, metadata);
            }
        }
    }
}
